plugin/activitywatch.vim

```diff
@@ -1,89 +1,136 @@
 endfunc
 
-function! s:CheckStatus()
-        if has('nvim')
-                if s:vimwatcher_job
-                        return 1
-                endif
-        elseif ch_status(s:vimwatcher_job) == "open"
-                return 1
-        endif
-        return 0
+function! s:HTTPPostOnStderr(jobid, data, event)
+    if a:data != ['']
+        let l:jobid_str = printf('%d', a:jobid)
+        let s:http_response_code[l:jobid_str] = a:data
+        "echo printf('aw-watcher-vim job %d stderr: %s', a:jobid, a:data)
+    endif
 endfunc
 
-function! s:StopVimWatcher()
-        if has('nvim')
-                call jobstop(s:vimwatcher_job)
-                let s:vimwatcher_job = 0
-        else
-                call job_stop(s:vimwatcher_job)
-        endif
+function! s:CreateBucket()
+    let l:body = {
+        \ 'name': s:bucketname,
+        \ 'hostname': hostname(),
+        \ 'client': 'aw-watcher-vim',
+        \ 'type': 'app.editor'
+    \}
+    call s:HTTPPostJson(s:bucket_apiurl, l:body)
 endfunc
 
-function! s:Send(msg)
-        if s:CheckStatus()
-                let l:json_msg = json_encode(a:msg)
-                if has('nvim')
-                        call jobsend(s:vimwatcher_job, l:json_msg . "\n")
-                else
-                        call ch_sendraw(s:vimwatcher_job, l:json_msg . "\n")
-                endif
-        endif
+function! s:Heartbeat()
+    " Only send heartbeats if we can connect to aw-server
+    if s:connected < 1
+        return
+    endif
+    let l:duration = 0
+    let l:localtime = localtime()
+    let l:timestamp = strftime('%FT%H:%M:%S%z')
+    let l:file = expand('%p')
+    let l:language = &filetype
+    let l:project = getcwd()
+    " Only send heartbeat if data was changed or more than 1 second has passed
+    " since last heartbeat
+    if    s:file != l:file ||
+        \ s:language != l:language ||
+        \ s:project != l:project ||
+        \ l:localtime - s:last_heartbeat > 1
+
+        let l:req_body = {
+            \ 'duration': 0,
+            \ 'timestamp': l:timestamp,
+            \ 'data': {
+                \ 'file': l:file,
+                \ 'language': l:language,
+                \ 'project': l:project
+            \ }
+        \}
+        call s:HTTPPostJson(s:heartbeat_apiurl, l:req_body)
+        let s:file = l:file
+        let s:language = l:language
+        let s:project = l:project
+        let s:last_heartbeat = l:localtime
+    endif
 endfunc
 
-function! s:Poke()
-        call s:Send({
-        \   'action': 'update',
-        \   'data': {
-        \       'file': expand('%:p'),
-        \       'language': &filetype,
-        \       'project': getcwd()
-        \}})
+function! AWStart()
+    call s:CreateBucket()
 endfunc
 
-call s:StartVimWatcher()
+function! AWStop()
+    let s:connected = 0
+endfunc
 
 augroup ActivityWatch
-        autocmd BufEnter * call s:StartVimWatcher()
-        autocmd CursorMoved,CursorMovedI * call s:Poke()
```

Comment:
Wouldn't this lead to extremely frequent heartbeats?

Reply:
In the old version aw-client pre-merged heartbeats.

Now in the new version I do things differently. Since we now have pre-merge heartbeats in aw-server and aw-server-rust it is fine to send heartbeats frequently, but as this would create a lot of HTTP requests I simply drop the heartbeat if it has the same data and re-send heartbeats with the same data once every second. The only difference in the final data in aw-server is that the resolution is 1 second instead of perfect precision in the old version with aw-client and 1s precision should be fine IMO.